NKL/Encode.lean

```diff
@@ -0,0 +1,331 @@
+-- TODO: if the nat is too large, it will be silently truncated.
+
+private def encNat (n : Nat) : ByteArray := .mk
+  #[ UInt8.ofNat $ n >>> 0
+   , UInt8.ofNat $ n >>> 8
+   , UInt8.ofNat $ n >>> 16
+   , UInt8.ofNat $ n >>> 24
+   , UInt8.ofNat $ n >>> 32
+   , UInt8.ofNat $ n >>> 40
+   , UInt8.ofNat $ n >>> 48
+   , UInt8.ofNat $ n >>> 56
+   ]
+
+private def decNat : DecodeM Nat := do
+  let nat8 : DecodeM Nat := next >>= fun x => return x.toNat
+  return (<- nat8) <<< 0  |||
+         (<- nat8) <<< 8  |||
+         (<- nat8) <<< 16 |||
+         (<- nat8) <<< 24 |||
+         (<- nat8) <<< 32 |||
+         (<- nat8) <<< 40 |||
+         (<- nat8) <<< 48 |||
+         (<- nat8) <<< 56
+
+#guard decode' decNat (.mk #[1,2,3,4,5,6,7]) == none
+#guard decode' decNat (.mk #[0,1,0,0,0,0,0,0]) == some 256
+#guard decode' decNat (encNat 0) == some 0
+#guard decode' decNat (encNat 1) == some 1
+#guard decode' decNat (encNat 256) == some 256
+
+------------------------------------------------------------------------------
+-- Ints are encoded as a sign tag followed by a Nat
+-- TODO: Int64 should be in standard library soon, revisit when available
+
+private def tag (t : UInt8) : List ByteArray -> ByteArray :=
+  List.foldl ByteArray.append (.mk #[t])
+
+private def encInt (i : Int) : ByteArray :=
+  match i with
+  | .ofNat n => tag 0 [encNat n]
+  | .negSucc n => tag 1 [encNat n]
+
+private def decInt : DecodeM Int := do
+  let sign <- next
+  let n <- decNat
+  bif sign == 0 then
+    return (.ofNat n)
+  else
+    return (.negSucc n)
+
+private def decInt' : DecodeM (Bool × Nat) := do
+  let sign <- next
+  let n <- decNat
+  return (sign == 1, n + sign.toNat)
+
+local instance : BEq ByteArray where
+  beq a b := a.data == b.data
+
+#guard encInt 1    == .mk #[0, 1, 0, 0, 0, 0, 0, 0, 0]
+#guard encInt 0    == .mk #[0, 0, 0, 0, 0, 0, 0, 0, 0]
+#guard encInt (-1) == .mk #[1, 0, 0, 0, 0, 0, 0, 0, 0]
+#guard encInt (-2) == .mk #[1, 1, 0, 0, 0, 0, 0, 0, 0]
+
+#guard decode' decInt (encInt 1) == some 1
+#guard decode' decInt (encInt 0) == some 0
+#guard decode' decInt (encInt $ -1) == some (-1)
+
+#guard decode' decInt' (encInt 1) == some (false, 1)
+#guard decode' decInt' (encInt 0) == some (false, 0)
+#guard decode' decInt' (encInt (-0)) == some (false, 0)
+#guard decode' decInt' (encInt $ -1) == some (true, 1)
+
+------------------------------------------------------------------------------
+-- Float values are encoded as (mantissa, exponent), where the
+-- mantissa and exponenet are integers.
+-- TODO:
+-- I am keeping the encode functions pure; encFloat can fail,
+-- but this seems unlikely as the Lean parser would have to not understand
+-- the result of toString on a Float value
+
+private def encFloat (f : Float) : ByteArray :=
+  let str := (toString f).toSubstring
+  let one : String.Pos := { byteIdx := 1 }
+  let sgn := str.get 0 == '-'
+  let str := bif sgn then { str with startPos := one } else str
```

Comment:
Why `bif` instead of `if`?

Reply:
I was just following something I read in the lean manual. `bif` is specialized to boolean conditions; not sure if there is really any difference.